src/core/vectortile/qgsvectortilemvtdecoder.cpp

```diff
@@ -101,280 +123,286 @@ QgsVectorTileFeatures QgsVectorTileMVTDecoder::layerFeatures( const QMap<QString
+      QHash<int, int> tagKeyIndexToFieldIndex;
+      for ( int i = 0; i < layer.keys_size(); ++i )
       {
-        // There is no assigned ID, but some parts of QGIS do not work correctly if all IDs are zero
-        // (e.g. labeling will not register two features with the same FID within a single layer),
-        // so let's generate some pseudo-unique FIDs to keep those bits happy
-        fid = featureNum;
-        fid |= ( layerNum & 0xff ) << 24;
-        fid |= ( static_cast<QgsFeatureId>( mTileID.row() ) & 0xff ) << 32;
-        fid |= ( static_cast<QgsFeatureId>( mTileID.column() ) & 0xff ) << 40;
+        const int fieldIndex = layerFields.indexOf( layer.keys( i ).c_str() );
+        if ( fieldIndex != -1 )
+          tagKeyIndexToFieldIndex.insert( i, fieldIndex );
       }
 
-      QgsFeature f( layerFields, fid );
-
-      //
-      // parse attributes
-      //
-
-      for ( int tagNum = 0; tagNum + 1 < feature.tags_size(); tagNum += 2 )
+      // go through features of a layer
+      for ( int featureNum = 0; featureNum < layer.features_size(); featureNum++ )
       {
-        const int keyIndex = static_cast<int>( feature.tags( tagNum ) );
-        const int fieldIndex = tagKeyIndexToFieldIndex.value( keyIndex, -1 );
-        if ( fieldIndex == -1 )
-          continue;
+        const ::vector_tile::Tile_Feature &feature = layer.features( featureNum );
 
-        const int valueIndex = static_cast<int>( feature.tags( tagNum + 1 ) );
-        if ( valueIndex >= layer.values_size() )
-        {
-          QgsDebugError( QStringLiteral( "Invalid value index for attribute" ) );
-          continue;
-        }
-        const ::vector_tile::Tile_Value &value = layer.values( valueIndex );
-
-        if ( value.has_string_value() )
-          f.setAttribute( fieldIndex, QString::fromStdString( value.string_value() ) );
-        else if ( value.has_float_value() )
-          f.setAttribute( fieldIndex, static_cast<double>( value.float_value() ) );
-        else if ( value.has_double_value() )
-          f.setAttribute( fieldIndex, value.double_value() );
-        else if ( value.has_int_value() )
-          f.setAttribute( fieldIndex, static_cast<int>( value.int_value() ) );
-        else if ( value.has_uint_value() )
-          f.setAttribute( fieldIndex, static_cast<int>( value.uint_value() ) );
-        else if ( value.has_sint_value() )
-          f.setAttribute( fieldIndex, static_cast<int>( value.sint_value() ) );
-        else if ( value.has_bool_value() )
-          f.setAttribute( fieldIndex, static_cast<bool>( value.bool_value() ) );
+        QgsFeatureId fid;
+#if 0
+        // even if a feature has an internal ID, it's not guaranteed to be unique across different
+        // tiles. This may violate the specifications, but it's been seen on mbtiles files in the wild...
+        if ( feature.has_id() )
+          fid = static_cast<QgsFeatureId>( feature.id() );
         else
+#endif
```

Comment:
Should we just remove this block altogether? Or Re-enable it the tiles size == 1 (i.e. when we only have one source)?

Reply:
@3nids any feedback on this one?